be/src/olap/storage_engine.cpp

```diff
@@ -117,8 +119,179 @@ StorageEngine::~StorageEngine() {
+        return true;
+    };
+    OLAPStatus clean_unfinished_rowset_meta_status = RowsetMetaManager::traverse_rowset_metas(data_dir->get_meta(), 
+        clean_unifinished_rowset_meta_func);
+    if (clean_unfinished_rowset_meta_status != OLAP_SUCCESS) {
+        // If failed to clean meta just skip the error, there will be useless metas in rocksdb column family
+        LOG(WARNING) << "there is failure when clean temp rowset meta from data dir:" << data_dir->path();
+    } else {
+        LOG(INFO) << "successfully clean temp rowset meta from data dir: " << data_dir->path();
+    }
+    return OLAP_SUCCESS;
+}
+
+OLAPStatus StorageEngine::_remove_old_meta_and_files(DataDir* data_dir) {
+    // clean old meta(olap header message) 
+    auto clean_old_meta_func = [this, data_dir](long tablet_id,
+        long schema_hash, const std::string& value) -> bool {
+        TabletMetaManager::remove(data_dir, tablet_id, schema_hash, OLD_HEADER_PREFIX);
+        LOG(INFO) << "successfully clean old tablet meta(olap header) for tablet=" 
+                  << tablet_id << "." << schema_hash 
+                  << "from data dir: " << data_dir->path();
+        return true;
+    };
+    OLAPStatus clean_old_meta_status = TabletMetaManager::traverse_headers(data_dir->get_meta(), 
+        clean_old_meta_func, OLD_HEADER_PREFIX);
+    if (clean_old_meta_status != OLAP_SUCCESS) {
+        // If failed to clean meta just skip the error, there will be useless metas in rocksdb column family
+        LOG(WARNING) << "there is failure when clean old tablet meta(olap header) from data dir:" << data_dir->path();
+    } else {
+        LOG(INFO) << "successfully clean old tablet meta(olap header) from data dir: " << data_dir->path();
+    }
+
+    // clean old files because they have hard links in new file name format
+    auto clean_old_files_func = [this, data_dir](long tablet_id,
+        long schema_hash, const std::string& value) -> bool {
+        TabletMetaPB tablet_meta_pb;
+        bool parsed = tablet_meta_pb.ParseFromString(value);
+        if (!parsed) {
+            // if errors when load, just skip it
+            LOG(WARNING) << "failed to load tablet meta from meta store to tablet=" << tablet_id << "." << schema_hash;
+            return true;
+        }
+
+        TabletSchema tablet_schema;
+        tablet_schema.init_from_pb(tablet_meta_pb.schema());
+        string data_path_prefix = data_dir->get_absolute_tablet_path(&tablet_meta_pb, true);
+
+        // convert visible pdelta file to rowsets and remove old files
+        for (auto& visible_rowset : tablet_meta_pb.rs_metas()) {
+            RowsetMetaSharedPtr alpha_rowset_meta(new AlphaRowsetMeta());
+            alpha_rowset_meta->init_from_pb(visible_rowset);
+            AlphaRowset rowset(&tablet_schema, data_path_prefix, data_dir, alpha_rowset_meta);
+            std::vector<std::string> old_files;
+            rowset.remove_old_files(&old_files);
```

Comment:
old_files is useless? why need it?

Reply:
because rowset remove method have a parameter to return the old files and segment group also return the files to remove.